localstack-core/localstack/services/dynamodbstreams/v2/provider.py

```diff
@@ -0,0 +1,74 @@
+import logging
+
+from localstack.aws import handlers
+from localstack.aws.api import RequestContext, ServiceRequest, ServiceResponse, handler
+from localstack.aws.api.dynamodbstreams import (
+    DescribeStreamInput,
+    DescribeStreamOutput,
+    DynamodbstreamsApi,
+    GetRecordsInput,
+    GetRecordsOutput,
+    GetShardIteratorInput,
+    GetShardIteratorOutput,
+    ListStreamsInput,
+    ListStreamsOutput,
+)
+from localstack.services.dynamodb.server import DynamodbServer
+from localstack.services.dynamodb.utils import modify_ddblocal_arns
+from localstack.services.dynamodb.v2.provider import DynamoDBProvider
+from localstack.services.plugins import ServiceLifecycleHook
+from localstack.utils.aws.arns import parse_arn
+
+LOG = logging.getLogger(__name__)
+
+
+class DynamoDBStreamsProvider(DynamodbstreamsApi, ServiceLifecycleHook):
+    def __init__(self):
+        self.server = DynamodbServer.get()
+
+    def on_after_init(self):
+        # add response processor specific to ddblocal
+        handlers.modify_service_response.append(self.service, modify_ddblocal_arns)
+
+    def forward_request(
+        self, context: RequestContext, service_request: ServiceRequest = None
+    ) -> ServiceResponse:
+        """
+        Forward a request to DynamoDB Local.
+        """
+        DynamoDBProvider.prepare_request_headers(
+            context.request.headers, account_id=context.account_id, region_name=context.region
+        )
+        return self.server.proxy(context, service_request)
+
+    def modify_stream_arn_for_ddb_local(self, stream_arn: str) -> str:
+        parsed_arn = parse_arn(stream_arn)
+
+        return f"arn:aws:dynamodb:ddblocal:000000000000:{parsed_arn['resource']}"
+
+    @handler("DescribeStream", expand=False)
+    def describe_stream(
+        self,
+        context: RequestContext,
+        payload: DescribeStreamInput,
+    ) -> DescribeStreamOutput:
+        payload["StreamArn"] = self.modify_stream_arn_for_ddb_local(payload.get("StreamArn", ""))
+        return self.forward_request(context, payload)
+
+    @handler("GetRecords", expand=False)
+    def get_records(self, context: RequestContext, payload: GetRecordsInput) -> GetRecordsOutput:
+        payload["ShardIterator"] = self.modify_stream_arn_for_ddb_local(
+            payload.get("ShardIterator", "")
+        )
+        return self.forward_request(context, payload)
+
+    @handler("GetShardIterator", expand=False)
+    def get_shard_iterator(
+        self, context: RequestContext, payload: GetShardIteratorInput
+    ) -> GetShardIteratorOutput:
+        payload["StreamArn"] = self.modify_stream_arn_for_ddb_local(payload.get("StreamArn", ""))
```

Comment:
Just a note: I'm not sure if modifying the request `payload` directly has side effects on the handler chain? At least for the request `context` we take precautions to restore the original state.

Reply:
Right yes, I normally never do that! Good catch!